File: Mage/Map/Cache/CacheOverlays.m
//
//  CacheOverlays.m
//  MAGE
//
//  Created by Brian Osborn on 12/17/15.
//  Copyright © 2015 National Geospatial Intelligence Agency. All rights reserved.
//

#import "CacheOverlays.h"
#import "CacheOverlay.h"
#import "CacheOverlayListener.h"
#import "GeoPackageCacheOverlay.h"
#import "MAGE-Swift.h"

@interface CacheOverlays ()

@property (nonatomic, strong) NSMutableDictionary<NSString *, CacheOverlay *> * overlays;
@property (nonatomic, strong) NSMutableArray<NSString *> * overlayNames;
@property (nonatomic, strong) NSMutableArray<NSObject<CacheOverlayListener> *> * listeners;
@property (nonatomic, strong) NSMutableSet<NSString *> * processing;

@end

@implementation CacheOverlays

static CacheOverlays * instance;

+(CacheOverlays *) getInstance{
    if(instance == nil){
        instance = [[CacheOverlays alloc] init];
    }
    return instance;
}

-(instancetype) init{
    self = [super init];
    if(self){
        self.overlays = [[NSMutableDictionary alloc] init];
        self.overlayNames = [[NSMutableArray alloc] init];
        self.listeners = [[NSMutableArray alloc] init];
        self.processing = [[NSMutableSet alloc] init];
    }
    return self;
}

-(void) registerListener: (NSObject<CacheOverlayListener> *) listener{
    @synchronized(self) {
        [self.listeners addObject:listener];
        [listener cacheOverlaysUpdated:[self getOverlays]];
    }
}

-(void) unregisterListener: (NSObject<CacheOverlayListener> *) listener{
    @synchronized(self) {
        [self.listeners removeObject:listener];
    }
}

-(void) setCacheOverlays:(NSArray<CacheOverlay *> *)overlays{
    @synchronized(self) {
        _overlays = [[NSMutableDictionary alloc] init];
        _overlayNames = [[NSMutableArray alloc] init];
        [self addCacheOverlays:overlays];
    }
}

-(void) addCacheOverlays:(NSArray<CacheOverlay *> *)overlays{
    @synchronized(self) {
        for(CacheOverlay * overlay in overlays){
            [self addCacheOverlayHelper:overlay];
        }
        [self notifyListeners];
    }
}

-(void) addCacheOverlay:(CacheOverlay *)overlay{
    @synchronized(self) {
        [self addCacheOverlayHelper:overlay];
        [self notifyListeners];
    }
}

-(void) addCacheOverlayHelper:(CacheOverlay *)overlay{
    NSString * cacheName = [overlay getCacheName];
    CacheOverlay * existingOverlay = [_overlays objectForKey:cacheName];
    if(existingOverlay == nil){
        [_overlayNames addObject:cacheName];
    }else{
        // Set existing cache overlays to their current enabled state
        [overlay setEnabled:existingOverlay.enabled];
        // If a new version of an existing cache overlay was added
        if(overlay.added){
            // Set the cache overlay being replaced
            if(existingOverlay.replacedCacheOverlay != nil){
                [overlay setReplacedCacheOverlay:existingOverlay.replacedCacheOverlay];
            }else{
                [overlay setReplacedCacheOverlay:existingOverlay];
            }
        }
    }
    [_overlays setObject:overlay forKey:cacheName];
}

-(void) notifyListeners{
    @synchronized(self) {
        [self notifyListenersExceptCaller:nil];
    }
}

-(void) notifyListenersExceptCaller:(NSObject<CacheOverlayListener> *) caller{
    @synchronized(self) {
        for(NSObject<CacheOverlayListener> * listener in self.listeners){
            if(caller == nil || listener != caller){
                [listener cacheOverlaysUpdated:[self getOverlays]];
            }
        }
    }
}

-(NSArray<CacheOverlay *> *) getOverlays{
    NSMutableArray<CacheOverlay *> *overlaysInCurrentEvent = [[NSMutableArray alloc] init];
    
    
    for(CacheOverlay * cacheOverlay in [self.overlays allValues]) {
        if ([cacheOverlay isKindOfClass:[GeoPackageCacheOverlay class]]) {
            GeoPackageCacheOverlay *gpCacheOverlay = (GeoPackageCacheOverlay *)cacheOverlay;
            NSString *filePath = gpCacheOverlay.filePath;
            // check if this filePath is consistent with a downloaded layer and if so, verify that layer is in this event
            NSArray *pathComponents = [filePath pathComponents];
            if ([[pathComponents objectAtIndex:[pathComponents count] - 3] isEqualToString:@"geopackages"]) {
                NSString *layerId = [pathComponents objectAtIndex:[pathComponents count] - 2];
                // check if this layer is in the event
                NSUInteger count = [Layer MR_countOfEntitiesWithPredicate:[NSPredicate predicateWithFormat:@"eventId == %@ AND remoteId == %ld", [Server currentEventId], layerId.integerValue] inContext:[NSManagedObjectContext MR_defaultContext]];
                if (count != 0) {
                    [overlaysInCurrentEvent addObject:cacheOverlay];
                }
            } else {
                [overlaysInCurrentEvent addObject:cacheOverlay];
            }
        } else {
            [overlaysInCurrentEvent addObject:cacheOverlay];
        }
    }
    return overlaysInCurrentEvent;
}

-(NSArray <CacheOverlay *> *) getLocallyLoadedOverlays {
    NSMutableArray<CacheOverlay *> *localOverlays = [[NSMutableArray alloc] init];
    
    for(CacheOverlay * cacheOverlay in [self.overlays allValues]) {
        if ([cacheOverlay isKindOfClass:[GeoPackageCacheOverlay class]]) {
            GeoPackageCacheOverlay *gpCacheOverlay = (GeoPackageCacheOverlay *)cacheOverlay;
            NSString *filePath = gpCacheOverlay.filePath;
            // check if this filePath is consistent with a downloaded layer and if so, verify that layer is in this event
            NSArray *pathComponents = [filePath pathComponents];
            if (![[pathComponents objectAtIndex:[pathComponents count] - 3] isEqualToString:@"geopackages"]) {
                [localOverlays addObject:cacheOverlay];
            }
        } else {
            [localOverlays addObject:cacheOverlay];
        }
    }
    return localOverlays;
}

-(NSUInteger) count{
    return [self.overlayNames count];
}

-(CacheOverlay *) atIndex:(NSUInteger)index{
    return [self.overlays objectForKey:[self.overlayNames objectAtIndex:index]];
}

-(CacheOverlay *) getByCacheName: (NSString *) cacheName{
    return [self.overlays objectForKey:cacheName];
}

-(void) removeCacheOverlay: (CacheOverlay *) overlay{
    [self removeByCacheName:[overlay getCacheName]];
}

-(void) removeByCacheName: (NSString *) cacheName{
    @synchronized(self) {
        [self.overlays removeObjectForKey:cacheName];
        [self.overlayNames removeObject:cacheName];
        [self notifyListeners];
    }
}

-(void) addProcessing: (NSString *) name{
    @synchronized(self) {
        [self.processing addObject:name];
        [self notifyListeners];
    }
}

-(void) addProcessingFromArray: (NSArray *) names{
    @synchronized(self) {
        [self.processing addObjectsFromArray:names];
        [self notifyListeners];
    }
}

-(void) removeProcessing: (NSString *) name{
    @synchronized(self) {
        [self.processing removeObject:name];
        [self notifyListeners];
    }
}

-(NSArray *) getProcessing{
    return [self.processing allObjects];
}

@end

File: Mage/Map/Cache/CacheOverlays.swift
//
//  CacheOverlays.m
//  MAGE
//
//  Created by Brian Osborn on 12/17/15.
//  Copyright © 2015 National Geospatial Intelligence Agency. All rights reserved.
//

import Foundation

@objc protocol CacheOverlayListener: NSObjectProtocol {
    @objc func cacheOverlaysUpdated(_ cacheOverlays: [CacheOverlay])
}

// TODO: This should be an actor
actor CacheOverlays: NSObject {
    @Injected(\.layerRepository)
    var layerRepository: LayerRepository
    
    static let shared = CacheOverlays()
    
    var overlays: [String: CacheOverlay] = [:]
    var overlayNames: [String] = []
    var listeners: [CacheOverlayListener] = []
    var processing: [String] = []
    
    @objc static func getInstance() -> CacheOverlays {
        shared
    }
    
    @objc func register(_ listener: CacheOverlayListener) async {
        listeners.append(listener)
        listener.cacheOverlaysUpdated(await getOverlays())
    }
    
    func unregisterListener(_ listener: CacheOverlayListener) {
        listeners.removeAll(where: { $0 === listener })
    }
    
    func setCacheOverlays(overlays: [CacheOverlay]) async {
        self.overlays = [:]
        self.overlayNames = []
        await add(overlays)
    }
    
    func add(_ overlays: [CacheOverlay]) async {
        for overlay in overlays {
            addCacheOverlayHelper(overlay: overlay)
        }
        await notifyListeners()
    }
    
    func addCacheOverlayHelper(overlay: CacheOverlay) {
        let cacheName = overlay.name
        if let existingOverlay = overlays[cacheName] {
            overlay.enabled = existingOverlay.enabled

            if overlay.added {
                overlay.replaced = existingOverlay.replaced ?? existingOverlay
            }
        } else {
            overlayNames.append(cacheName)
        }
        overlays[cacheName] = overlay
    }
    
    func addCacheOverlay(overlay: CacheOverlay) async {
        addCacheOverlayHelper(overlay: overlay)
        await notifyListeners()
    }
    
    @objc func notifyListeners() async {
        await notifyListenersExceptCaller(caller: nil)
    }
    
    @objc func notifyListenersExceptCaller(caller: (any CacheOverlayListener)?) async {
        let overlaySnapshot = await getOverlays()
        
        for listener in listeners {
            if caller == nil || !listener.isEqual(caller) {
                listener.cacheOverlaysUpdated(overlaySnapshot)
            }
        }
    }
    
    @objc func getOverlays() async -> [CacheOverlay] {
        var overlaysInCurrentEvent: [CacheOverlay] = []
        
        let sortedNames = overlayNames.sorted()
        
        for name in sortedNames {
            guard let overlay = overlays[name] else { continue }
            
            if await shouldIncludeOverlay(overlay) {
                overlaysInCurrentEvent.append(overlay)
            }
        }
        
        return overlaysInCurrentEvent
    }
    
    private func shouldIncludeOverlay(_ overlay: CacheOverlay) async -> Bool {
        // If it's a GeoPackage overlay, validate it against the current event
        if let geoOverlay = overlay as? GeoPackageCacheOverlay,
           let layerId = geoOverlay.layerId,
           let layerIdInt = Int(layerId),
           let currentEventId = Server.currentEventId()
        {
            let count = await layerRepository.count(eventId: currentEventId, layerId: layerIdInt)
            return count != 0
        }

        // All non-GeoPackage overlays are always included
        return true
    }
    
    func count() -> Int {
        overlayNames.count
    }
    
    func atIndex(index: Int) -> CacheOverlay? {
        overlays[overlayNames[index]]
    }
    
    func getByCacheName(_ cacheName: String?) -> CacheOverlay? {
        guard let cacheName = cacheName else { return nil }
        return overlays[cacheName]
    }
    
    @objc func removeCacheOverlay(overlay: CacheOverlay) async {
        await remove(byCacheName: overlay.cacheName)
    }
    
    func remove(byCacheName: String) async {
        overlays.removeValue(forKey: byCacheName)
        overlayNames.removeAll(where: { $0 == byCacheName })
        await notifyListeners()
    }
    
    func addProcessing(name: String) async {
        self.processing.append(name)
        await notifyListeners()
    }
    
    func addProcessing(from: [String]?) async {
        self.processing.append(contentsOf: from ?? [])
        await notifyListeners()
    }
    
    func removeProcessing(_ name: String) async {
        self.processing.removeAll(where: { $0 == name })
        await notifyListeners()
    }
    
    func getProcessing() -> [String] {
        processing
    }
    
    func removeAll() async {
        for overlay in overlays.values {
            await removeCacheOverlay(overlay: overlay)
        }
    }
}
